cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,1076 @@
+  using store_t = typename container_t::mapped_type;
+  container_t edge_type_to_kv_store;
+};
+
+template <typename edge_id_t, typename edge_type_t, typename value_t>
+search_container_t<edge_id_t, edge_type_t, value_t>::~search_container_t()
+{
+  pimpl.reset();
+}
+
+template <typename edge_id_t, typename edge_type_t, typename value_t>
+search_container_t<edge_id_t, edge_type_t, value_t>::search_container_t()
+  : pimpl{std::make_unique<search_container_impl<edge_id_t, edge_type_t, value_t>>()}
+{
+}
+
+template <typename edge_id_t, typename edge_type_t, typename value_t>
+search_container_t<edge_id_t, edge_type_t, value_t>::search_container_t(
+  raft::handle_t const& handle, std::vector<edge_type_t> typs, std::vector<edge_id_t> typ_freqs)
+  : pimpl{std::make_unique<search_container_impl<edge_id_t, edge_type_t, value_t>>(
+      handle, typs, typ_freqs)}
+{
+}
+
+template <typename edge_id_t, typename edge_type_t, typename value_t>
+search_container_t<edge_id_t, edge_type_t, value_t>::search_container_t(const search_container_t&)
+{
+}
+
+template <typename edge_id_t, typename edge_type_t, typename value_t>
+void search_container_t<edge_id_t, edge_type_t, value_t>::insert(
+  raft::handle_t const& handle,
+  edge_type_t type,
+  raft::device_span<edge_id_t const> edge_ids_to_insert,
+  decltype(cugraph::allocate_dataframe_buffer<value_t>(0,
+                                                       rmm::cuda_stream_view{}))&& values_to_insert)
+{
+  pimpl->insert(handle, type, edge_ids_to_insert, std::move(values_to_insert));
+}
+
+template <typename edge_id_t, typename edge_type_t, typename value_t>
+std::optional<decltype(cugraph::allocate_dataframe_buffer<value_t>(0, rmm::cuda_stream_view{}))>
+search_container_t<edge_id_t, edge_type_t, value_t>::src_dst_from_edge_id_and_type(
+  raft::handle_t const& handle,
+  raft::device_span<edge_id_t const> edge_ids_to_lookup,
+  edge_type_t edge_type_to_lookup,
+  bool multi_gpu) const
+{
+  return pimpl->src_dst_from_edge_id_and_type(
+    handle, edge_ids_to_lookup, edge_type_to_lookup, multi_gpu);
+}
+
+template <typename edge_id_t, typename edge_type_t, typename value_t>
+std::optional<decltype(cugraph::allocate_dataframe_buffer<value_t>(0, rmm::cuda_stream_view{}))>
+search_container_t<edge_id_t, edge_type_t, value_t>::src_dst_from_edge_id_and_type(
+  raft::handle_t const& handle,
+  raft::device_span<edge_id_t const> edge_ids_to_lookup,
+  raft::device_span<edge_type_t const> edge_types_to_lookup,
+  bool multi_gpu) const
+{
+  return pimpl->src_dst_from_edge_id_and_type(
+    handle, edge_ids_to_lookup, edge_types_to_lookup, multi_gpu);
+}
+
+namespace detail {
+
+template <typename GraphViewType,
+          typename EdgeIdInputWrapper,
+          typename EdgeTypeInputWrapper,
+          typename EdgeTypeAndIdToSrcDstLookupContainerType>
+EdgeTypeAndIdToSrcDstLookupContainerType build_edge_id_and_type_to_src_dst_lookup_map(
```

Comment:
This might be a good case for a specialization for multi_gpu == true and a separate specialization for multi_gpu == false.  The logic in this function is hard to follow with respect to what is the flow in SG and what is the flow in MG.

Reply:
Simplified.